### PR TITLE
Add generic field types

### PR DESCRIPTION
**Pull request by codecholeric**
This next step to fully support Generics within ArchUnit adds support for generic field types. In particular

* `JavaField.getType()` now returns a `JavaParameterizedType` for a parameterized generic field type and the raw types otherwise
* all type arguments of generic field types are added to `JavaClass.directDependencies{From/To}Self`

Example: ArchUnit would now detect `String` as a `Dependency` of a field declaration like

```
class SomeClass {
    Set<List<? super String>> someField;
}
```

### CODE REVIEW

File: archunit/src/test/java/com/tngtech/archunit/core/importer/ClassFileImporterGenericFieldTypesTest.java [hunk 1]
```diff
@@ -0,0 +1,522 @@
...
+    }
+
+    @Test
+    public void imports_generic_field_type_with_one_type_argument() {
+        @SuppressWarnings("unused")
+        class GenericFieldType<T> {
+        }
+        @SuppressWarnings("unused")
+        class SomeClass {
+            GenericFieldType<String> field;
```
Comment (hankem) [inline archunit/src/test/java/com/tngtech/archunit/core/importer/ClassFileImporterGenericFieldTypesTest.java:52]: If this was `GenericFieldType<String>[] field`, what would `genericFieldType` be like?
I'd have expected that its erasure was `GenericFieldType[].class`, but this doesn't seem to be the case.

----
By the way, for a `NonGenericFieldType[] field` in the test above,
```
 assertThatType(fieldType).as("generic field type").hasErasure(NonGenericFieldType[].class)
```
works, but
```
assertThatType(fieldType).as("generic field type").matches(NonGenericFieldType[].class);
```
doesn't.

**Review by hankem — CHANGES_REQUESTED** [review]
It seems that there is still something off with
* generic array fields, and
* fields with primitive arrays as generic type arguments

(but I couldn't find a solution yet).

File: archunit/src/main/java/com/tngtech/archunit/core/importer/SignatureTypeArgumentProcessor.java [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+            this.builder = builder;
+        }
+
+        @Override
+        public JavaType finish(OWNER owner, Iterable<JavaTypeVariable<?>> allTypeParametersInContext, ClassesByTypeName classes) {
+            return builder.build(owner, allTypeParametersInContext, classes);
+        }
+    }
+
+    static class ReferenceCreationProcess<OWNER extends HasDescription> implements DomainBuilders.JavaTypeCreationProcess<OWNER> {
```
[the remark this answers is not in the document]
Comment (codecholeric) [reply archunit/src/main/java/com/tngtech/archunit/core/importer/SignatureTypeArgumentProcessor.java:182]: Yeah, agree, don't know why I used the qualified type names here :joy:

File: archunit/src/main/java/com/tngtech/archunit/core/domain/Dependency.java [hunk 3]
```diff
@@ -145,16 +145,19 @@ static Dependency fromInheritance(JavaClass origin, JavaClass targetSupertype) {
...
 
     static Set<Dependency> tryCreateFromGenericInterfaceTypeArgument(JavaClass originClass, JavaType genericInterface, JavaClass typeArgumentDependency) {
-        return createGenericDependency(originClass, "interface", genericInterface, typeArgumentDependency);
+        return tryCreateDependency(originClass, genericDependencyType("interface", genericInterface), typeArgumentDependency);
     }
 
-    private static Set<Dependency> createGenericDependency(JavaClass originClass, String genericTypeDescription, JavaType genericSuperType, JavaClass typeArgumentDependency) {
-        String dependencyType = "has generic " + genericTypeDescription + " " + bracketFormat(genericSuperType.getName()) + " with type argument depending on";
-        return tryCreateDependency(originClass, originClass.getDescription(), dependencyType, typeArgumentDependency, originClass.getSourceCodeLocation());
+    static Set<Dependency> tryCreateFromGenericFieldTypeArgument(JavaField origin, JavaClass typeArgumentDependency) {
```
[the remark this answers is not in the document]
Comment (hankem) [reply archunit/src/main/java/com/tngtech/archunit/core/domain/Dependency.java:155]: Neat! ♥ Believe it or not, but I haven't looked at that branch yet...

File: archunit/src/test/java/com/tngtech/archunit/core/importer/ClassFileImporterGenericFieldTypesTest.java [hunk 4]
```diff
@@ -0,0 +1,522 @@
...
+        }
+
+        JavaType fieldType = new ClassFileImporter().importClass(SomeClass.class).getField("field").getType();
+
+        assertThatType(fieldType).as("generic field type").matches(NonGenericFieldType.class);
+    }
+
+    @Test
+    public void imports_generic_field_type_with_one_type_argument() {
+        @SuppressWarnings("unused")
```
[the remark this answers is not in the document]
Comment (codecholeric) [reply archunit/src/test/java/com/tngtech/archunit/core/importer/ClassFileImporterGenericFieldTypesTest.java:47]: The `@SuppressWarnings` here simply covers the case of the unused type parameter. That's why `NonGenericFieldType` doesn't need it. From my point of view within ArchUnit I would generally suppress all warnings of local classes of `@Test` methods :joy: 
But what I think makes sense is to not have any warnings in valid code :man_shrugging: If those warnings don't make sense for our case, because the info is dug up from the byte code and thus still valid, we should suppress them, no?

File: archunit/src/main/java/com/tngtech/archunit/core/domain/Dependency.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (codecholeric) [reply archunit/src/main/java/com/tngtech/archunit/core/domain/Dependency.java:155]: better so :joy:

File: archunit/src/test/java/com/tngtech/archunit/core/importer/ClassFileImporterGenericFieldTypesTest.java (hunk 1 above)
Comment (codecholeric) [reply archunit/src/test/java/com/tngtech/archunit/core/importer/ClassFileImporterGenericFieldTypesTest.java:52]: For the first part: Yes, that might be an error in `toErasure()` :thinking: 
For the second part, this is likely, because `Foo[]` can't be imported from any class file. So maybe there is a discrepancy in the behavior for missing type creation for array types with respect to the Reflection API :thinking:

**Comment by codecholeric** [comment]
Okay, this was more tedious than I would have hoped, but I think I got most array type bugs now :wink: Can you check again? Thanks a lot!

File: archunit/src/test/java/com/tngtech/archunit/core/domain/JavaClassTest.java [hunk 5]
```diff
@@ -544,28 +552,69 @@ public void direct_dependencies_from_self_by_inheritance_through_generic_interfa
...
+                        .withExpectedDescriptionTemplate("has generic interface <" + genericInterfaceSignature + "> with type argument depending on <#target>")
                         .to(Comparable.class, Map.class, Map.Entry.class, String.class, BufferedInputStream[][].class, Serializable.class, List.class, Set.class, Iterable.class, File.class)
 
                         .from(Child.class).to(BufferedInputStream.class).inLocation(getClass(), 0)
                         .withDescriptionContaining("depends on component type <%s>", BufferedInputStream.class.getName())
                 );
     }
 
+    @Test
+    public void direct_dependencies_from_self_by_field_declarations() {
```
Comment (hankem) [inline archunit/src/test/java/com/tngtech/archunit/core/domain/JavaClassTest.java:577]: Should `SomeClass` have a dependency on `File` in the following example?
```java
        class Outer<T extends File> {
            class SomeClass {
                List<? extends T> field;
            }
        }
```

Comment (hankem) [reply archunit/src/test/java/com/tngtech/archunit/core/domain/JavaClassTest.java:577]: I first thought we could extend `JavaClassDependencies.dependenciesOfTypes` with
```
        } else if (javaType instanceof JavaTypeVariable<?>) {
            result.addAll(dependenciesOfTypes(((JavaTypeVariable<?>) javaType).getBounds()));
```
but that gives me some `StackOverflowError`s... 

I haven't actually looked into them, but dut does this mean that we're missing dependencies from some type's erasure?

**Review by hankem — APPROVED** [review]
Looks very good!
Apart from a few minor comments, I've only found one potentially missing dependency due to a `JavaTypeVariable`.

**Comment by hankem** [comment]
Not required for this PR, but a question for a further improvement:
Would you expect that `javaClass.getAllFields()` resolves types of parameterized generic superclasses? I.e. should this pass:
```java
@Test
public void field_from_generic_base_class() {
    class B<T> {
        T field;
    }
    class C extends B<Long> {
    }
    JavaClass javaClass = new ClassFileImporter().importClasses(B.class, C.class, Long.class).get(C.class);
    JavaField field = javaClass.getAllFields().stream().filter(f -> f.getName().equals("field")).findAny().get();
    assertThatType(field.getRawType()).matches(Long.class);  // is currently Object.class
    assertThatType(field.getType()).matches(Long.class);  // is currently JavaTypeVariable{T}
}
```

File: archunit/src/main/java/com/tngtech/archunit/core/domain/Formatters.java [hunk 6]
```diff
@@ -122,4 +123,26 @@ public static String ensureSimpleName(String name) {
...
+     * Returns the canonical array type name of any array type name passed in. Otherwise
+     * returns the passed type name as is. For example {@code [Ljava.lang.String;} will
+     * be reformatted to {@code java.lang.String[]} or {@code [I} will be reformatted
+     * to {@code int[]}, while {@code java.lang.String} would simply be returned as is.
+     * @param typeName A Java type name
+     * @return the passed type name, but for array type names the canonical array type name
+     */
+    @PublicAPI(usage = ACCESS)
+    public static String ensureCanonicalArrayTypeName(String typeName) {
+        if (!typeName.startsWith("[")) {
```
[the remark this answers is not in the document]
Comment (codecholeric) [reply archunit/src/main/java/com/tngtech/archunit/core/domain/Formatters.java:136]: I'll extract it to a function, but I think we shouldn't unify it with `JavaClassDescriptor.From.isArray`, because that method is more widely used and supports array class names as well as canonical names. But if we already have a canonical name, we don't really have to do anything. I mean, it wouldn't hurt, but I feel the overhead and "publification" of the private method `isArray` is not worth it :man_shrugging: Let's follow the rule of three, if we run into it a 3rd time we'll reconsider :wink:

File: archunit/src/main/java/com/tngtech/archunit/core/domain/JavaWildcardType.java [hunk 7]
```diff
@@ -58,7 +57,7 @@
     @Override
     @PublicAPI(usage = ACCESS)
     public String getName() {
-        return WILDCARD_TYPE_NAME;
+        return WILDCARD_TYPE_NAME + boundsToString();
```
[the remark this answers is not in the document]
Comment (codecholeric) [reply archunit/src/main/java/com/tngtech/archunit/core/domain/JavaWildcardType.java:60]: True and done :+1:

File: archunit/src/jdk9test/java/com/tngtech/archunit/core/domain/JavaParameterizedTypeTest.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        class WithWildcards<TEST extends List<Map<Map<?, ? extends File>, ? super Map<? extends String[][], ? extends int[][]>>>> {
+        }
+        return $$(
+                createTestInput(WithConcreteTypeName.class),
+                createTestInput(WithTypeVariable.class),
+                createTestInput(WithNestedType.class),
+                createTestInput(WithArrays.class),
+                createTestInput(WithTypeVariableArrays.class),
+                createTestInput(WithWildcards.class)
+        );
```
[the remark this answers is not in the document]
Comment (codecholeric) [reply archunit/src/jdk9test/java/com/tngtech/archunit/core/domain/JavaParameterizedTypeTest.java:45]: Well, I've been trapped in the Java 7 world with ArchUnit for too long :joy: I'll adjust it :+1:

File: archunit/src/test/java/com/tngtech/archunit/core/domain/JavaClassTest.java [hunk 9]
```diff
@@ -519,12 +519,20 @@ public void direct_dependencies_from_self_by_inheritance_through_generic_supercl
 
         JavaClass javaClass = importClassWithContext(Child.class);
 
+        String genericSuperclass = Base.class.getName() + "<"
```
[the remark this answers is not in the document]
Comment (codecholeric) [reply archunit/src/test/java/com/tngtech/archunit/core/domain/JavaClassTest.java:522]: Yeah, that's kind of true :thinking: I'll relax the description assertion, because in the end it's only used to make sure we have one dependency per target type. The exact description is not the issue here, you're right, so we shouldn't assert it too strictly.

File: archunit-integration-test/src/test/java/com/tngtech/archunit/testutils/ExpectedDependency.java [hunk 10]
```diff
@@ -29,12 +31,16 @@ public static TypeParameterCreator typeParameter(Class<?> clazz, String typePara
...
     }
 
-    public static GenericSupertypeTypeArgumentCreator genericSuperclass(Class<?> clazz, Class<?> genericSuperclassErasure) {
-        return new GenericSupertypeTypeArgumentCreator(clazz, "superclass", genericSuperclassErasure);
+    public static GenericSupertypeTypeArgumentCreator genericSuperclassOf(Class<?> clazz) {
+        return new GenericSupertypeTypeArgumentCreator(clazz, "superclass", clazz.getGenericSuperclass());
+    }
+
+    public static GenericSupertypeTypeArgumentCreator genericInterfaceOf(Class<?> clazz) {
+        return new GenericSupertypeTypeArgumentCreator(clazz, "interface", clazz.getGenericInterfaces()[0]);
```
[the remark this answers is not in the document]
Comment (codecholeric) [reply archunit-integration-test/src/test/java/com/tngtech/archunit/testutils/ExpectedDependency.java:39]: Yeah, probably better. I mean, I thought the singular of the method name strongly suggests that it only works correctly with one interface :joy: But yes, better to make these assertions crystal clear and avoid some "fake tests" in the future that overlook another interface :+1:

File: archunit/src/test/java/com/tngtech/archunit/core/importer/ClassFileImporterGenericFieldTypesTest.java [hunk 11]
```diff
@@ -0,0 +1,614 @@
...
+        }
+
+        JavaType genericFieldType = new ClassFileImporter().importClasses(SomeClass.class, String.class)
+                .get(SomeClass.class).getField("field").getType();
+
+        assertThatType(genericFieldType).as("generic field type")
+                .hasErasure(GenericFieldType.class)
+                .hasActualTypeArguments(String.class);
+    }
+
```
[the remark this answers is not in the document]
Comment (codecholeric) [reply archunit/src/test/java/com/tngtech/archunit/core/importer/ClassFileImporterGenericFieldTypesTest.java:64]: Yeah, I kind of think you shouldn't use smelly raw types as field types, but still good to document the case :wink:

[the remark this answers is not in the document]
Comment (codecholeric) [reply archunit/src/test/java/com/tngtech/archunit/core/importer/ClassFileImporterGenericFieldTypesTest.java:64]: Also added it to superclasses / interfaces, since it's valid there as well IMHO...

File: archunit/src/test/java/com/tngtech/archunit/core/domain/JavaClassTest.java (hunk 5 above)
Comment (codecholeric) [reply archunit/src/test/java/com/tngtech/archunit/core/domain/JavaClassTest.java:577]: Hmm, that's an interesting case :thinking: I guess since `SomeClass` can use the bound of `T` the bound must be a dependency...

Comment (codecholeric) [reply archunit/src/test/java/com/tngtech/archunit/core/domain/JavaClassTest.java:577]: The danger of `StackOverflowError` is not that surprising, since you can define `class Foo<T extends Foo<T>>` :wink:

Comment (codecholeric) [reply archunit/src/test/java/com/tngtech/archunit/core/domain/JavaClassTest.java:577]: I pondered about it for a while and I think it's actually okay that the inner class does not report the bound of the outer class defined type variable as a dependency per se. Because if the inner class never makes use of that bound, then you could argue that it does not really depend on the bound (you could throw away the bound on the outer type variable and nothing would change for the inner class). However if the inner class would make use of the bound (in your example operate on `File`), then there would be some other dependency on that type present and it would behave reasonably again.

Comment (codecholeric) [reply archunit/src/test/java/com/tngtech/archunit/core/domain/JavaClassTest.java:577]: If there should be many complaints about this corner case we can still revisit the issue :wink:

**Comment by codecholeric** [comment]
> Not required for this PR, but a question for a further improvement:
> Would you expect that `javaClass.getAllFields()` resolves types of parameterized generic superclasses? I.e. should this pass:
> 
> ```java
> @Test
> public void field_from_generic_base_class() {
>     class B<T> {
>         T field;
>     }
>     class C extends B<Long> {
>     }
>     JavaClass javaClass = new ClassFileImporter().importClasses(B.class, C.class, Long.class).get(C.class);
>     JavaField field = javaClass.getAllFields().stream().filter(f -> f.getName().equals("field")).findAny().get();
>     assertThatType(field.getRawType()).matches(Long.class);  // is currently Object.class
>     assertThatType(field.getType()).matches(Long.class);  // is currently JavaTypeVariable{T}
> }
> ```

At the moment `getAllFields()` behaves kind of consistently with the Reflection API. The Reflection API also does not provide any easy way to resolve the superclass actual type arguments. (that's why classes like https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/core/GenericTypeResolver.html exist)
We can see if anybody actually misses this within ArchUnit. Because regarding dependencies I would argue again, that members from the superclass have no dependencies on types only defined by the parameterization of the subclass.

**Comment by codecholeric** [comment]
Thanks so much for all the effort you put into the review @hankem :smiley: I really appreciate it!

File: archunit/src/test/java/com/tngtech/archunit/core/domain/JavaClassTest.java (hunk 5 above)
Comment (hankem) [reply archunit/src/test/java/com/tngtech/archunit/core/domain/JavaClassTest.java:566]: Sounds totally fair!
